p2p/src/authenticated/lookup/actors/spawner/actor.rs

```diff
@@ -0,0 +1,142 @@
+        context.register(
+            "messages_received",
+            "messages received",
+            received_messages.clone(),
+        );
+        context.register(
+            "messages_rate_limited",
+            "messages rate limited",
+            rate_limited.clone(),
+        );
+        let (sender, receiver) = mpsc::channel(cfg.mailbox_size);
+
+        (
+            Self {
+                context,
+                mailbox_size: cfg.mailbox_size,
+                ping_frequency: cfg.ping_frequency,
+                allowed_ping_rate: cfg.allowed_ping_rate,
+                receiver,
+                connections,
+                sent_messages,
+                received_messages,
+                rate_limited,
+            },
+            Mailbox::new(sender),
+        )
+    }
+
+    pub fn start(
+        mut self,
+        tracker: tracker::Mailbox<E, C>,
+        router: router::Mailbox<C>,
+    ) -> Handle<()> {
+        self.context.spawn_ref()(self.run(tracker, router))
+    }
+
+    async fn run(mut self, tracker: tracker::Mailbox<E, C>, router: router::Mailbox<C>) {
+        while let Some(msg) = self.receiver.next().await {
+            match msg {
+                Message::Spawn {
+                    peer,
+                    connection,
+                    reservation,
+                } => {
+                    // Mark peer as connected
+                    self.connections.inc();
+
+                    // Clone required variables
+                    let connections = self.connections.clone();
+                    let sent_messages = self.sent_messages.clone();
+                    let received_messages = self.received_messages.clone();
+                    let rate_limited = self.rate_limited.clone();
+                    let tracker = tracker.clone();
+                    let mut router = router.clone();
+
+                    // Spawn peer
+                    self.context
+                        .with_label("peer")
+                        .spawn(move |context| async move {
+                            // Create peer
+                            debug!(?peer, "peer started");
+                            let (actor, messenger) = peer::Actor::new(
+                                context,
+                                peer::Config {
+                                    ping_frequency: self.ping_frequency,
+                                    allowed_ping_rate: self.allowed_ping_rate,
+                                    sent_messages,
+                                    received_messages,
+                                    rate_limited,
+                                    mailbox_size: self.mailbox_size,
+                                },
+                                reservation,
+                            );
+
+                            // Register peer with the router
+                            let channels = router.ready(peer.clone(), messenger).await;
```

Comment:
This isn't related to your changes, but I think it would be clearer to actually move the `tracker.connect()` call here (for both `discovery` and `lookup`).

They both functionally do the same thing (in fact we could avoid passing in the reservation to the peer for the same reason) if we want to drop near the router release below.